### PR TITLE
add v3 interfaces and adjust mgmt funcs

### PR DESCRIPTION
**Pull request by michaelcohen716**

### CODE REVIEW

File: contracts/xKNC.sol [hunk 1]
```diff
@@ -210,14 +216,58 @@ contract xKNC is
...
+     */
+    function claimReward(
+        uint256 cycle,
+        uint256 index,
+        IERC20[] calldata tokens,
+        uint256[] calldata cumulativeAmounts,
+        bytes32[] calldata merkleProof,
+        uint256[] calldata minRates
+    ) external onlyOwnerOrManager {
+        require(tokens.length == cumulativeAmounts.length, "Must be equal length");
```
Comment (Anyhowclick) [inline contracts/xKNC.sol:249]: this check is also done in the `claim` function. maybe can remove to save gas

File: contracts/interface/IRewardsDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+interface IRewardsDistributor {
+  function claim(
+    uint256 cycle,
+    uint256 index,
+    address user,
+    IERC20[] calldata tokens,
+    uint256[] calldata cumulativeAmounts,
+    bytes32[] calldata merkleProof
+  ) external returns (uint256[] memory claimAmounts);
+}
```
Comment (Anyhowclick) [inline contracts/interface/IRewardsDistributor.sol:12]: missing newline

File: contracts/xKNC.sol [hunk 3]
```diff
@@ -210,14 +216,58 @@ contract xKNC is
...
+        rewardsDistributor.claim(
+            cycle,
+            index,
+            address(this),
+            tokens,
+            cumulativeAmounts,
+            merkleProof
+        );
+
+        for (uint256 i = 0; i < tokens.length; i++) {
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply contracts/xKNC.sol:261]: Suggest to allow upgrade KyberProxy address for flexible usage

File: contracts/xKNC.sol [hunk 4]
```diff
@@ -61,6 +62,11 @@ contract xKNC is
 
     enum FeeTypes {MINT, BURN, CLAIM}
 
+    // vars added for v3
+    bool private v3Initialized;
+    IRewardsDistributor private rewardsDistributor;
```
Comment (ducquangkstn) [inline contracts/xKNC.sol:67]: Pls make this upgradable and add a getter for it

File: contracts/xKNC.sol [hunk 5]
```diff
@@ -513,4 +472,31 @@ contract xKNC is
     receive() external payable {
         require(msg.sender != tx.origin, "Errant ETH deposit");
     }
+
+    // 
+    // V3 MIGRATION
+    // 
+    // 1) pause mint/burn 2) unstake all 3) approve migrator 4) migrate to new token 5) execute v3Initialize function with new addresses for new contracts, 6) restake
+
+    function migrateV3(
```
Comment (ducquangkstn) [inline contracts/xKNC.sol:481]: owner or mangers can create a fake staking address to withdraw funds.

It is nicer if kyberDao, knc and kyberStaking is immutable. So the values of these variable is specified when creating new implementation

Comment (michaelcohen716) [reply contracts/xKNC.sol:481]: `migrateV3` function can only be called once so dao, knc and staking addresses cannot be changed in this implementation

Comment (ducquangkstn) [reply contracts/xKNC.sol:434]: Agreed. It should be ok if you trust mangers that they do not font-run you when doing migration

File: contracts/xKNC.sol [hunk 6]
```diff
@@ -210,124 +217,61 @@ contract xKNC is
...
-
-            if (kyberFeeTokens[i] == ETH_ADDRESS) {
-                _administerEthFee(FeeTypes.CLAIM, ethBalBefore);
+        for (uint256 i = 0; i < tokens.length; i++) {
+            if(address(tokens[i]) == address(knc)){
+                continue;
+            }else if(address(tokens[i]) == ETH_ADDRESS){
+                _swapEtherToKnc(getFundEthBalanceWei(), minRates[i]);
+            } else {
+                _swapTokenToKnc(address(tokens[i]), tokens[i].balanceOf(address(this)), minRates[i]);
```
Comment (ducquangkstn) [inline contracts/xKNC.sol:267]: there is a chance that `tokens[i].balanceOf(address(this)) == 0`

Comment (Anyhowclick) [reply contracts/xKNC.sol:267]: same for ether

**Pull request merged**